README
pod2text Geo::Google::StaticMaps::V2.pm > README

If this is still here it means the programmer was too lazy to create the readme file.

You can create it now by using the command shown above from this directory.

At the very least you should be able to use this set of instructions
to install the module...

perl Makefile.PL
make
make test
make install

If you are on a windows box you should use 'nmake' rather than 'make'.


Changes
Revision history for Perl module Geo::Google::StaticMaps::V2

0.09 2013-06-16
  - Updated packages to support Perl < 5.10

0.08-2 2013-06-13
  - Updated Documentation

0.08 2013-06-13
  - Updated path->color, path->fillcolor, and marker->color methods
    - Added color format support [$r,$g,$b],
    - Added color format support {r=>$r,g=>$g,b=>$b}
    - Added path color alpha support "10%" or 0-255
    - Added tests

0.07 2013-06-13
  - Updated map->save method to set format (localized!) based on file extension unless defined
  - Added map->visible method and refactored the Path and Markers packages

0.06 2013-06-13
  - Updated documentation

0.05 2013-06-12
  - Added Google Maps API Signature capability

0.04 2013-06-10
  - Fixed Makefile.PL requires
  - Added RPM spec file
  - Added t/006_save.t

0.03 2013-06-09
  - Path method and object are now fully operational

0.02 2013-06-09
  - Marker method and object are now fully operational

0.01 2013-06-09
  - original version; created by ExtUtils::ModuleMaker 0.51


lib/Geo/Google/StaticMaps/V2.pm
package Geo::Google::StaticMaps::V2;
use warnings;
use strict;
use base qw{Package::New};
use File::Basename qw{};
use Path::Class qw{file};
use URI qw{};
use LWP::UserAgent qw{};
use Geo::Google::StaticMaps::V2::Markers;
use Geo::Google::StaticMaps::V2::Path;

our $VERSION = '0.09';
our $PACKAGE = __PACKAGE__;

=head1 NAME

Geo::Google::StaticMaps::V2 - Generate Images from Google Static Maps V2 API

=head1 SYNOPSIS

  use Geo::Google::StaticMaps::V2;
  my $map=Geo::Google::StaticMaps::V2->new;
  print $map->url;
  print $map->image;
  $map->save("filename.png");

=head1 DESCRIPTION

The packages generates images from the Google Static Maps V2 API which can be saved locally for use in accordance with your license with Google.

=head1 USAGE

=head1 CONSTRUCTORS

=head2 new

  use Geo::Google::StaticMaps::V2;
  my $map=Geo::Google::StaticMaps::V2->new(
                                           width    => 600,
                                           height   => 480,
                                           sensor   => 0,
                                           scale    => 1,
                                           format   => "png8",
                                           type     => "roadmap",
                                           protocol => "http",
                                           server   => "maps.googleapis.com",
                                           script   => "/maps/api/staticmap",
                                          );

Any property can be specified on construction but all have sane defaults and are not required to be set.

=head2 marker

Creates a L<Geo::Google::StaticMaps::V2::Markers> object and adds the object to the internal Markers array.

  $map->marker(location=>"7140 Main Street, Clifton, Virginia 20124");
  $map->marker(location=>{lat=>38.780676,lon=>-77.387105});
  $map->marker(location=>[38.780676,-77.387105]);
  $map->marker(location=>"38.780676,-77.387105");

markers (optional) define one or more markers to attach to the image at specified locations. Multiple markers may be placed within the same markers parameter as long as they exhibit the same style; you may add additional markers of differing styles by adding additional markers parameters. Note that if you supply markers for a map, you do not need to specify the (normally required) center and zoom parameters.

Examples

  my $marker1=$map->marker(location=>{lat=>38.780676,lon=>-77.387105}); #isa L<Geo::Google::StaticMaps::V2::Markers>
  $marker1->addLocation([38.780513,-77.387128]);                        #second point shares style with first point

  my $marker2=$map->marker(locations=>[
                                       {lat=>38.780596,lon=>-77.386837},
                                       [38.780346,-77.386923]
                                      ]);                               #third and forth points with different style
  $marker2->size("tiny");                                               #third and forth points are now tiny
  $marker1->color("blue");                                              #first and second point are now blue

=cut

sub marker {
  my $self=shift;
  my $obj=Geo::Google::StaticMaps::V2::Markers->new(@_);
  push @{$self->_markers}, $obj;
  return $obj;
}

#head2 _markers
#
#Returns an array reference with the current list of marker objects.
#
#cut

sub _markers {
  my $self=shift;
  $self->{"_markers"}=[] unless ref($self->{"_markers"}) eq "ARRAY";
  return $self->{"_markers"};
}

=head2 path

Creates a L<Geo::Google::StaticMaps::V2::Path> object and adds the object to the internal Paths array.

path (optional) defines a single path of two or more connected points to overlay on the image at specified locations. Note that if you supply a path for a map, you do not need to specify the (normally required) center and zoom parameters.

=cut

sub path {
  my $self=shift;
  my $obj=Geo::Google::StaticMaps::V2::Path->new(@_);
  push @{$self->_paths}, $obj;
  return $obj;
}

#head2 _paths
#
#Returns an array reference with the current list of path objects.
#
#cut

sub _paths {
  my $self=shift;
  $self->{"_paths"}=[] unless ref($self->{"_paths"}) eq "ARRAY";
  return $self->{"_paths"};
}

=head2 visible

Creates a L<Geo::Google::StaticMaps::V2::Visible> object and adds the object to the internal Visibles array.

visible (optional) specifies one or more locations that should remain visible on the map, though no markers or other indicators will be displayed. Use this parameter to ensure that certain features or map locations are shown on the static map.

=cut

sub visible {
  my $self=shift;
  my $obj=Geo::Google::StaticMaps::V2::Visible->new(@_);
  push @{$self->_visibles}, $obj;
  return $obj;
}

#head2 _visibles
#
#Returns an array reference with the current list of visible objects.
#
##cut

sub _visibles {
  my $self=shift;
  $self->{"_visibles"}=[] unless ref($self->{"_visibles"}) eq "ARRAY";
  return $self->{"_visibles"};
}

=head1 PROPERTIES

=head2 center

center (required if markers not present) defines the center of the map, equidistant from all edges of the map. This parameter takes a location as either a comma-separated {latitude,longitude} pair (e.g. "40.714728,-73.998672") or a string address (e.g. "City Hall, New York, NY") identifying a unique location on the face of the earth.

=cut

sub center {
  my $self=shift;
  $self->{"center"}=shift if @_;
  return $self->{"center"};
}

=head2 zoom

zoom (required if markers not present) defines the zoom level of the map, which determines the magnification level of the map. This parameter takes a numerical value corresponding to the zoom level of the region desired.

=cut

sub zoom {
  my $self=shift;
  $self->{"zoom"}=shift if @_;
  return $self->{"zoom"};
}

=head2 width

Width part of size parameter. 

UOM: pixels

Note: width of image is actually width times scale

=cut

sub width {
  my $self=shift;
  $self->{"width"}=shift if @_;
  $self->{"width"}||=600;
  return $self->{"width"};
}

=head2 height

Height part of size parameter

UOM: pixels

Note: height of image is actually height times scale

=cut

sub height {
  my $self=shift;
  $self->{"height"}=shift if @_;
  $self->{"height"}||=400;
  return $self->{"height"};
}

=head2 sensor

Sets or returns the sensor value true or false setting is Perlish.

  $map->sensor(0);       #default
  $map->sensor("false"); #Do not do this as "false" is true to Perl

sensor (required) specifies whether the application requesting the static map is using a sensor to determine the user's location. This parameter is required for all static map requests.

=cut

sub sensor {
  my $self=shift;
  $self->{"sensor"}=shift if @_;
  $self->{"sensor"}||=0;
  return $self->{"sensor"};
}

=head2 scale

scale (optional) affects the number of pixels that are returned. scale=2 returns twice as many pixels as scale=1 while retaining the same coverage area and level of detail (i.e. the contents of the map don't change). This is useful when developing for high-resolution displays, or when generating a map for printing. The default value is 1. Accepted values are 2 and 4 (4 is only available to Maps API for Business customers.) See Scale Values for more information. 

  $map->scale; #undef (default is 1; not passed on URL), 1, 2, 4

=cut

sub scale {
  my $self=shift;
  $self->{"scale"}=shift if @_;
  return $self->{"scale"};
}

=head2 format

format (optional) defines the format of the resulting image. By default, the Static Maps API creates PNG images. There are several possible formats including GIF, JPEG and PNG types. Which format you use depends on how you intend to present the image. JPEG typically provides greater compression, while GIF and PNG provide greater detail. For more information, see Image Formats. 

  $map->format; #undef (default is png8; not passed on URL), png8, png32, gif, jpg, jpg-baseline

=cut

sub format {
  my $self=shift;
  $self->{"format"}=shift if @_;
  return $self->{"format"};
}

=head2 type

maptype (optional) defines the type of map to construct. There are several possible maptype values, including roadmap, satellite, hybrid, and terrain.

  $map->type; #undef (default is roadmap; not passed on URL), roadmap, satellite, terrain, hybrid

=cut

sub type {
  my $self=shift;
  $self->{"type"}=shift if @_;
  return $self->{"type"};
}

=head2 server

Sets or returns the Google Maps API server

  $map->server("maps.googleapis.com");  #default

=cut

sub server {
  my $self=shift;
  $self->{"server"}=shift if @_;
  $self->{"server"}="maps.googleapis.com" unless defined($self->{"server"});
  return $self->{"server"};
}

=head2 script

Sets or returns the script for the Google Maps API Static Map

  $map->script("/maps/api/staticmap");  #default

=cut

sub script {
  my $self=shift;
  $self->{"script"}=shift if @_;
  $self->{"script"}="/maps/api/staticmap" unless defined($self->{"script"});
  return $self->{"script"};
}

=head2 protocol

Sets or returns the protocol

  $map->protocol("http");  #default
  $map->protocol("https"); #https to avoid cross security domain issues in browsers but uses more resources

=cut

sub protocol {
  my $self=shift;
  $self->{"protocol"}=shift if @_;
  $self->{"protocol"}="http" unless defined($self->{"protocol"});
  return $self->{"protocol"};
}

#head2 _service
#
#Returns a cached URL with only "$protocol://$server"
#
#cut

sub _service {
  my $self=shift;
  $self->{"_service"}=shift if @_;
  unless (defined $self->{"_service"}) {
    $self->{"_service"}=URI->new;
    $self->{"_service"}->scheme($self->protocol);
    $self->{"_service"}->host($self->server);
  }
  return $self->{"_service"};
}

=head1 METHODS

=head2 image

Returns the image as a binary scalar.

  my $blob=$map->image;

See: L<LWP::UserAgent>

=cut

sub image {
  my $self=shift;
  my $response=$self->_ua->get($self->url);
  if ($response->is_success) {
    return $response->decoded_content;
  } else {
    die $response->status_line;
  }
}

#head2 _ua
#
#Returns a cached LWP::UserAgent object
#
#cut

sub _ua {
  my $self=shift;
  unless (defined $self->{"_ua"}) {
    $self->{"_ua"}=LWP::UserAgent->new;
    $self->{"_ua"}->timeout(10);
    $self->{"_ua"}->env_proxy;
  }
  return $self->{"_ua"};
}

=head2 save

Saves image to the local file system.

  $map->save("image.png");
  $map->save("image.gif");
  $map->save("image.jpg");

Note: If you have not explicitly set the format, the save method will guess the format from the extension.

=cut

sub save {
  my $self     = shift;
  my $filename = shift;
  my $suffix   = (File::Basename::fileparse($filename, keys(%{$self->_extensions})))[2];
  local $self->{"format"}=$self->_extensions->{$suffix}
    if (exists($self->_extensions->{$suffix}) && !defined($self->format));
  my $file=file($filename); #isa Path::Class::File
  my $fh=$file->openw;
  print $fh $self->image;
  return $self;
}

sub _extensions {
  return {".png" => undef, ".gif" => "gif", ".jpg" => "jpg"};
}

=head2 url

Returns the URL for the Static map.  If L<URL::Signature::Google::Maps::API> is installed and configured the URL is seamlessly signed with your Google Enterprise Key.

  my $url=$map->url; #isa L<URI>

=cut

sub url {
  my $self=shift;
  my $url=URI->new($self->script); #isa URI
  my @q=();
  push @q, size    => $self->_size;  #required
  push @q, maptype => $self->type   if defined $self->type;
  push @q, scale   => $self->scale  if defined $self->scale;
  push @q, format  => $self->format  if defined $self->format;
  push @q, sensor  => $self->sensor ? "true" : "false";
  my $needs_view=1;
  foreach my $visible (@{$self->_visibles}) {
    $needs_view=0;
    push @q, visible  => $visible->stringify;
  }
  foreach my $marker (@{$self->_markers}) {
    $needs_view=0;
    push @q, markers => $marker->stringify;
  }
  foreach my $path (@{$self->_paths}) {
    $needs_view=0;
    push @q, path => $path->stringify;
  }
  if (defined $self->center) {
    push @q, center => $self->center;
  } else {
    push @q, center => "Clifton, VA" if $needs_view; #we should only get here in the trival case
  }
  if (defined $self->zoom) {
    push @q, zoom => $self->zoom;
  } else {
    push @q, zoom => "7" if $needs_view;             #we should only get here in the trival case
  }
  $url->query_form(@q);
  if ($self->_signer) {
    $url=$self->_signer->url($self->_service => $url);
  } else {
    $url->scheme($self->protocol);
    $url->host($self->server);
  }
  return $url;
}

#head2 _size
#
#size (required) defines the rectangular dimensions of the map image. This parameter takes a string of the form {horizontal_value}x{vertical_value}. For example, 500x400 defines a map 500 pixels wide by 400 pixels high. Maps smaller than 180 pixels in width will display a reduced-size Google logo. This parameter is affected by the scale parameter, described below; the final output size is the product of the size and scale values. 
#
#cut

sub _size {
  my $self=shift;
  return join("x", $self->width, $self->height);
}

#head2 _signer
#
#Returns false but defined or a cached L<> object
#
#Note: This method returns false if the URL::Signature::Google::Maps::API is not installed and thus URL signatures will not be available.
#
#cut

sub _signer {
  my $self=shift;
  $self->{"_signer"}=shift if @_;
  unless (defined $self->{"_signer"}) { #init
    eval('use URL::Signature::Google::Maps::API');
    if ($@) {
      $self->{"_signer"}="";
    } else {
      $self->{"_signer"}=URL::Signature::Google::Maps::API->new(channel=>$self->channel, client=>$self->client, key=>$self->key)
        unless defined($self->{"_signer"});
    }
  }
  return $self->{"_signer"};
}

=head1 Google Enterprise Credentials

These settings are simply passed through to L<URL::Signature::Google::Maps::API>.

I recommend storing the credentials in the INI formatted file and leaving these values as undef.

=head2 client

Sets and returns the Google Enterprise Client

=cut

sub client {
  my $self=shift;
  $self->{"client"}=shift if @_;
  return $self->{"client"};
}

=head2 key

Sets and returns the Google Enterprise Key

=cut

sub key {
  my $self=shift;
  $self->{"key"}=shift if @_;
  return $self->{"key"};
}

=head2 channel

Sets and returns the Google Enterprise channel.

=cut

sub channel {
  my $self=shift;
  $self->{"channel"}=shift if @_;
  return $self->{"channel"};
}

=head1 BUGS

Please log on RT and send an email to the author.

=head1 SUPPORT

DavisNetworks.com supports all Perl applications including this package.

=head1 AUTHOR

  Michael R. Davis
  CPAN ID: MRDVT
  Satellite Tracking of People, LLC
  mdavis@stopllc.com
  http://www.stopllc.com/

=head1 COPYRIGHT

This program is free software licensed under the...

  The General Public License (GPL) Version 2, June 1991

The full text of the license can be found in the LICENSE file included with this module.

=head1 SEE ALSO

L<Geo::Google::StaticMaps>, L<Geo::Google::MapObject>, L<Net::Flickr::Geo::GoogleMaps>, L<URL::Signature::Google::Maps::API>

=cut

1;


lib/Geo/Google/StaticMaps/V2/Base.pm
package Geo::Google::StaticMaps::V2::Base;
use warnings;
use strict;
use base qw{Package::New};
our $VERSION = '0.03';
our $PACKAGE = __PACKAGE__;

=head1 NAME

Geo::Google::StaticMaps::V2::Base - Generate Images from Google Static Maps API V2 API

=head1 SYNOPSIS

  use Geo::Google::StaticMaps::V2;
  my $map=Geo::Google::StaticMaps::V2->new;
  my $marker=$map->marker(lacation=>[$lat=>$lon]); #isa Geo::Google::StaticMaps::V2::Base

=head1 DESCRIPTION

The packages generates images from the Google Static Maps API V2 API which can be saved locally for use in accordance with your license with Google.

=head1 USAGE


=head2 initialize

=cut

sub initialize {
  my $self=shift;
  %$self=@_;
  $self->addLocation(delete $self->{"location"}) if exists $self->{"location"};
}

=head2 stringify

=cut

sub stringify {
  my $self=shift;
  my @p=();
  push @p, $self->_styles;
  foreach my $location ($self->locations) {
    my $string="";
    if (ref($location) eq "HASH") {
      $string=sprintf("%0.6f,%0.6f", $location->{"lat"}, $location->{"lon"});
    } elsif (ref($location) eq "ARRAY") {
      $string=sprintf("%0.6f,%0.6f", @$location);
    } else {
      $string="$location"; #or any object that overloads
    }
    push @p, $string;
  }
  return join "|", @p;
}

=head2 locations

Returns the locations array which can be set upon construction.

=cut

sub locations {
  my $self=shift;
  $self->{"locations"}=[] unless ref($self->{"locations"}) eq "ARRAY";
  return wantarray ? @{$self->{"locations"}} : $self->{"locations"};
}

=head1 PROPERTIES

=head2 color

color: (optional) specifies a 24-bit color (example: color=0xFFFFCC) or a predefined color from the set {black, brown, green, purple, yellow, blue, gray, orange, red, white}.

=cut

sub color {
  my $self=shift;
  $self->{"color"}=shift if @_;
  return $self->{"color"};
}

=head1 METHODS

=head2 addLocation

  $marker->addLocation("Clifton, VA");

=cut

sub addLocation {
  my $self=shift;
  push @{$self->locations}, @_;
  return $self;
}

=head1 BUGS

=head1 SUPPORT

=head1 AUTHOR

  Michael R. Davis
  CPAN ID: MRDVT
  Satellite Tracking of People, LLC
  mdavis@stopllc.com
  http://www.stopllc.com/

=head1 COPYRIGHT

This program is free software licensed under the...

  The General Public License (GPL) Version 2, June 1991

The full text of the license can be found in the LICENSE file included with this module.

=head1 SEE ALSO

L<Geo::Google::StaticMaps>, L<Geo::Google::MapObject>, L<Net::Flickr::Geo::GoogleMaps>

=cut

1;


lib/Geo/Google/StaticMaps/V2/Path.pm
package Geo::Google::StaticMaps::V2::Path;
use warnings;
use strict;
use base qw{Geo::Google::StaticMaps::V2::Visible};
our $VERSION = '0.08';
our $PACKAGE = __PACKAGE__;

=head1 NAME

Geo::Google::StaticMaps::V2::Path - Generate Images from Google Static Maps V2 API

=head1 SYNOPSIS

  use Geo::Google::StaticMaps::V2;
  my $map=Geo::Google::StaticMaps::V2->new;
  my $path=$map->path(locations=>["Clifton, VA", "Pag, Croatia"], geodesic=>1); #isa Geo::Google::StaticMaps::V2::Path
  print $map->url, "\n";

=head1 DESCRIPTION

The packages generates images from the Google Static Maps V2 API which can be saved locally for use in accordance with your license with Google.

=head1 USAGE

=head1 PROPERTIES

=cut

sub _styles {
  my $self=shift;
  my @styles=();
  push @styles, join ":", weight    => $self->weight                      if defined $self->weight;
  push @styles, join ":", color     => $self->color                       if defined $self->color;
  push @styles, join ":", fillcolor => $self->fillcolor                   if defined $self->fillcolor;
  push @styles, join ":", geodesic  => $self->geodesic ? "true" : "false" if defined $self->geodesic;
  return @styles;
}

=head2 weight

weight: (optional) specifies the thickness of the path in pixels. If no weight parameter is set, the path will appear in its default thickness (5 pixels). 

=cut

sub weight {
  my $self=shift;
  $self->{"weight"}=shift if @_;
  return $self->{"weight"};
}

=head2 color

color: (optional) specifies a color either as a 24-bit (example: color=0xFFFFCC) or 32-bit hexadecimal value (example: color=0xFFFFCCFF), or from the set {black, brown, green, purple, yellow, blue, gray, orange, red, white}.

When a 32-bit hex value is specified, the last two characters specify the 8-bit alpha transparency value. This value varies between 00 (completely transparent) and FF (completely opaque). Note that transparencies are supported in paths, though they are not supported for markers.

  my $color=$path->color("blue");
  my $color=$path->color("0xFFFFCC");
  my $color=$path->color({r=>255,g=>0,b=>0,a=>64}); #maps to red   25% returns 0xFF000040
  my $color=$path->color([0,255,0,"75%"]);          #maps to green 75% returns 0x00FF00C0

=cut

sub _color {
  my $self=shift;
  my $key=shift; #color||fillcolor
  $self->{$key}=shift if @_;
  if (ref($self->{$key})) {
    my $r;
    my $g;
    my $b;
    my $a;
    if (ref($self->{$key}) eq "HASH") {
      $r = $self->{$key}->{"r"} || 0;
      $g = $self->{$key}->{"g"} || 0;
      $b = $self->{$key}->{"b"} || 0;
      $a = $self->{$key}->{"a"};
    } elsif (ref($self->{$key}) eq "ARRAY") {
      $r = $self->{$key}->[0]   || 0;
      $g = $self->{$key}->[1]   || 0;
      $b = $self->{$key}->[2]   || 0;
      $a = $self->{$key}->[3];
    } else {
      die("Error: Unknown reference type expecting HASH or ARRAY.");
    }
    if (defined $a) {
      if ($a =~ m/^(100|\d\d|\d)\%$/) {
        $a=int($1/100*255);
      }
      return sprintf("0x%02X%02X%02X%02X", $r, $g, $b, $a);
    } else {
      return sprintf("0x%02X%02X%02X", $r, $g, $b);
    }
  } else {
    return $self->{$key};
  }
}

sub color {shift->_color(color=>@_)};

=head2 fillcolor

fillcolor: (optional) indicates both that the path marks off a polygonal area and specifies the fill color to use as an overlay within that area. The set of locations following need not be a "closed" loop; the Static Map server will automatically join the first and last points. Note, however, that any stroke on the exterior of the filled area will not be closed unless you specifically provide the same beginning and end location. 

=cut

sub fillcolor {shift->_color(fillcolor=>@_)};

=head2 geodesic

geodesic: (optional) indicates that the requested path should be interpreted as a geodesic line that follows the curvature of the Earth. When false, the path is rendered as a straight line in screen space. Defaults to false.

=cut

sub geodesic {
  my $self=shift;
  $self->{"geodesic"}=shift if @_;
  return $self->{"geodesic"};
}

=head1 METHODS

=head2 addLocation

  $marker->addLocation("Clifton, VA");

=head1 BUGS

Please log on RT and send an email to the author.

=head1 SUPPORT

DavisNetworks.com supports all Perl applications including this package.

=head1 AUTHOR

  Michael R. Davis
  CPAN ID: MRDVT
  Satellite Tracking of People, LLC
  mdavis@stopllc.com
  http://www.stopllc.com/

=head1 COPYRIGHT

This program is free software licensed under the...

  The General Public License (GPL) Version 2, June 1991

The full text of the license can be found in the LICENSE file included with this module.

=head1 SEE ALSO

L<Geo::Google::StaticMaps::V2>

=cut

1;


lib/Geo/Google/StaticMaps/V2/Visible.pm
package Geo::Google::StaticMaps::V2::Visible;
use warnings;
use strict;
use base qw{Package::New};
our $VERSION = '0.07';
our $PACKAGE = __PACKAGE__;

=head1 NAME

Geo::Google::StaticMaps::V2::Visible - Generate Images from Google Static Maps V2 API

=head1 SYNOPSIS

  use Geo::Google::StaticMaps::V2;
  my $map=Geo::Google::StaticMaps::V2->new;
  my $visible=$map->visible(locations=>["Clifton, VA", "Pag, Croatia"]); #isa Geo::Google::StaticMaps::V2::Visible
  print $map->url, "\n";

=head1 DESCRIPTION

The packages generates images from the Google Static Maps V2 API which can be saved locally for use in accordance with your license with Google.

=head1 USAGE

  use Geo::Google::StaticMaps::V2;
  my $map=Geo::Google::StaticMaps::V2->new;
  my $visible=$map->visible(location=>"Clifton, VA"); #isa Geo::Google::StaticMaps::V2::Visible

=head2 initialize

Sets all passed in arguments and folds location parameter into the locations array.

=cut

sub initialize {
  my $self=shift;
  %$self=@_;
  $self->addLocation(delete $self->{"location"}) if exists $self->{"location"};
}

=head2 stringify

Handles various formats for locations seamlessly.

=cut

sub stringify {
  my $self=shift;
  my @p=();
  push @p, $self->_styles;
  foreach my $location ($self->locations) {
    my $string="";
    if (ref($location) eq "HASH") {
      $string=sprintf("%0.6f,%0.6f", $location->{"lat"}, $location->{"lon"});
    } elsif (ref($location) eq "ARRAY") {
      $string=sprintf("%0.6f,%0.6f", @$location);
    } else {
      $string="$location"; #or any object that overloads
    }
    push @p, $string;
  }
  return join "|", @p;
}

=head2 locations

Returns the locations array which can be set upon construction.

=cut

sub locations {
  my $self=shift;
  $self->{"locations"}=[] unless ref($self->{"locations"}) eq "ARRAY";
  return wantarray ? @{$self->{"locations"}} : $self->{"locations"};
}

sub _styles {
  my $self=shift;
  my @styles=();
  return @styles;
}

=head1 METHODS

=head2 addLocation

  $marker->addLocation("Clifton, VA");

=cut

sub addLocation {
  my $self=shift;
  push @{$self->locations}, @_;
  return $self;
}

=head1 BUGS

Please log on RT and send an email to the author.

=head1 SUPPORT

DavisNetworks.com supports all Perl applications including this package.

=head1 AUTHOR

  Michael R. Davis
  CPAN ID: MRDVT
  Satellite Tracking of People, LLC
  mdavis@stopllc.com
  http://www.stopllc.com/

=head1 COPYRIGHT

This program is free software licensed under the...

  The General Public License (GPL) Version 2, June 1991

The full text of the license can be found in the LICENSE file included with this module.

=head1 SEE ALSO

L<Geo::Google::StaticMaps::V2>, L<Geo::Google::StaticMaps::V2::Path>, L<Geo::Google::StaticMaps::V2::Markers>

=cut

1;


lib/Geo/Google/StaticMaps/V2/Markers.pm
package Geo::Google::StaticMaps::V2::Markers;
use warnings;
use strict;
use base qw{Geo::Google::StaticMaps::V2::Visible};
our $VERSION = '0.08';
our $PACKAGE = __PACKAGE__;

=head1 NAME

Geo::Google::StaticMaps::V2::Markers - Generate Images from Google Static Maps V2 API

=head1 SYNOPSIS

  use Geo::Google::StaticMaps::V2;
  my $map=Geo::Google::StaticMaps::V2->new;
  my $marker=$map->marker(location=>[$lat=>$lon]); #isa Geo::Google::StaticMaps::V2::Markers

=head1 DESCRIPTION

The packages generates images from the Google Static Maps V2 API which can be saved locally for use in accordance with your license with Google.

=head1 USAGE

=head1 PROPERTIES

=cut

sub _styles {
  my $self=shift;
  my @styles=();
  push @styles, join ":", size   => $self->size                      if defined $self->size;
  push @styles, join ":", color  => $self->color                     if defined $self->color;
  push @styles, join ":", label  => $self->label                     if defined $self->label;
  push @styles, join ":", icon   => $self->icon                      if defined $self->icon;
  push @styles, join ":", shadow => $self->shadow ? "true" : "false" if defined $self->shadow;
  return @styles;
}

=head2 size

size: (optional) specifies the size of marker from the set {tiny, mid, small}. If no size parameter is set, the marker will appear in its default (normal) size. 

=cut

sub size {
  my $self=shift;
  $self->{"size"}=shift if @_;
  return $self->{"size"};
}

=head2 color

Set and returns a formatted marker color code.

color: (optional) specifies a 24-bit color (example: color=0xFFFFCC) or a predefined color from the set {black, brown, green, purple, yellow, blue, gray, orange, red, white}.

  my $color=$marker->color("blue");
  my $color=$marker->color("0xFFFFCC");
  my $color=$marker->color({r=>255,g=>0,b=>0}); #maps to red
  my $color=$marker->color([255,0,0]);          #maps to red

=cut

sub color {
  my $self=shift;
  $self->{"color"}=shift if @_;
  if (ref($self->{"color"})) {
    my $r;
    my $g;
    my $b;
    if (ref($self->{"color"}) eq "HASH") {
      $r = $self->{"color"}->{"r"} || 0;
      $g = $self->{"color"}->{"g"} || 0;
      $b = $self->{"color"}->{"b"} || 0;
    } elsif (ref($self->{"color"}) eq "ARRAY") {
      $r = $self->{"color"}->[0]   || 0;
      $g = $self->{"color"}->[1]   || 0;
      $b = $self->{"color"}->[2]   || 0;
    } else {
      die("Error: Unknown reference type expecting HASH or ARRAY.");
    }
    return sprintf("0x%02X%02X%02X", $r, $g, $b);
  } else {
    return $self->{"color"};
  }
}

=head2 label

label: (optional) specifies a single uppercase alphanumeric character from the set {A-Z, 0-9}. (The requirement for uppercase characters is new to this version of the API.) Note that default and mid sized markers are the only markers capable of displaying an alphanumeric-character parameter. tiny and small markers are not capable of displaying an alphanumeric-character. 

=cut

sub label {
  my $self=shift;
  $self->{"label"}=shift if @_;
  return $self->{"label"};
}

=head2 icon

icon specifies a URL to use as the marker's custom icon. Images may be in PNG, JPEG or GIF formats, though PNG is recommended. 

The icon parameter must be specified using a URL (which should be URL-encoded). You may use any valid URL of your choosing, or a URL-shortening service such as http://goo.gl. Most URL-shortening services have the advantage of automatically encoding URLs. Icons are limited to sizes of 4096 pixels (64x64 for square images), and the Static Maps service allows up to five unique custom icons per request. Note that each of these unique icons may be used multiple times within the static map.

Custom icons that have a shadow:true descriptor (the default) will have their "anchor point" set as the bottom center of the provided icon image, from which the shadow is cast. Icons without a shadow (setting a shadow:false descriptor) are instead assumed to be icons centered on their specified locations, so their anchor points are set as the center of the image.

=cut

sub icon {
  my $self=shift;
  $self->{"icon"}=shift if @_;
  return $self->{"icon"};
}

=head2 shadow

shadow (default true) indicates that the Static Maps service should construct an appropriate shadow for the image. This shadow is based on the image's visible region and its opacity/transparency.

=cut

sub shadow {
  my $self=shift;
  $self->{"shadow"}=shift if @_;
  return $self->{"shadow"};
}

=head1 METHODS

=head2 addLocation

  $marker->addLocation("Clifton, VA");

=head1 BUGS

Please log on RT and send an email to the author.

=head1 SUPPORT

DavisNetworks.com supports all Perl applications including this package.

=head1 AUTHOR

  Michael R. Davis
  CPAN ID: MRDVT
  Satellite Tracking of People, LLC
  mdavis@stopllc.com
  http://www.stopllc.com/

=head1 COPYRIGHT

This program is free software licensed under the...

  The General Public License (GPL) Version 2, June 1991

The full text of the license can be found in the LICENSE file included with this module.

=head1 SEE ALSO

L<Geo::Google::StaticMaps::V2>

=cut

1;
